types/time.go

```diff
@@ -962,66 +959,63 @@ func ParseDuration(str string, fsp int) (Duration, error) {
 		str = str[n+1:]
 	}
 
-	var overflow bool
+	var (
+		integeralPart = str
+		fracPart      int
+		overflow      bool
+	)
 	if n := strings.IndexByte(str, '.'); n >= 0 {
 		// It has fractional precision parts.
 		fracStr := str[n+1:]
-		frac, overflow, err = ParseFrac(fracStr, fsp)
+		fracPart, overflow, err = ParseFrac(fracStr, fsp)
 		if err != nil {
 			return ZeroDuration, errors.Trace(err)
 		}
-		str = str[0:n]
+		integeralPart = str[0:n]
 	}
 
-	// It tries to split str with delimiter, time delimiter must be :
-	seps := strings.Split(str, ":")
+	// It tries to split integeralPart with delimiter, time delimiter must be :
+	seps := strings.Split(integeralPart, ":")
 
 	switch len(seps) {
 	case 1:
 		if dayExists {
 			hour, err = strconv.Atoi(seps[0])
 		} else {
 			// No delimiter.
-			if len(str) == 7 {
-				// HHHMMSS
-				_, err = fmt.Sscanf(str, "%3d%2d%2d", &hour, &minute, &second)
-			} else if len(str) == 6 {
-				// HHMMSS
-				_, err = fmt.Sscanf(str, "%2d%2d%2d", &hour, &minute, &second)
-			} else if len(str) == 5 {
-				// HMMSS
-				_, err = fmt.Sscanf(str, "%1d%2d%2d", &hour, &minute, &second)
-			} else if len(str) == 4 {
-				// MMSS
-				_, err = fmt.Sscanf(str, "%2d%2d", &minute, &second)
-			} else if len(str) == 3 {
-				// MSS
-				_, err = fmt.Sscanf(str, "%1d%2d", &minute, &second)
-			} else if len(str) == 2 {
-				// SS
-				_, err = fmt.Sscanf(str, "%2d", &second)
-			} else if len(str) == 1 {
-				// 0S
-				_, err = fmt.Sscanf(str, "%1d", &second)
-			} else {
-				// Maybe only contains date.
+			switch len(integeralPart) {
+			case 7: // HHHMMSS
```

Comment:
alternative way is using a map whose key is int and value is string. 

Suppose such may is `timeFmt`, then `timeFmt[7]` could just return `"%3d%2d%2d"`

Reply:
Yes, but it seems that every element in `timeFmt` will only be used in this function now.
I'm not sure whether it's better than using switch-case.

Reply:
map access is slower than a switch case, I prefer not to use map.